py/selenium/webdriver/common/options.py

```diff
@@ -22,117 +22,142 @@
 
-    @property
-    def page_load_strategy(self) -> str:
-        """
-        :returns: page load strategy if set, the default is "normal"
-        """
-        return self._caps["pageLoadStrategy"]
+class _UnHandledPromptBehavior:
+    """How the driver should respond when an alert is present and the:
+    command sent is not handling the alert:
+    https://w3c.github.io/webdriver/#dfn-table-of-page-load-strategies:
 
-    @page_load_strategy.setter
-    def page_load_strategy(self, strategy: str) -> None:
-        """Determines the point at which a navigation command is returned:
-        https://w3c.github.io/webdriver/#dfn-table-of-page-load-strategies.
+    :param behavior: behavior to use when an alert is encountered
 
-        :param strategy: the strategy corresponding to a document readiness state
-        """
-        if strategy in ["normal", "eager", "none"]:
-            self.set_capability("pageLoadStrategy", strategy)
-        else:
-            raise ValueError("Strategy can only be one of the following: normal, eager, none")
+    :returns: Values for implicit timeout, pageLoad timeout and script timeout if set (in milliseconds)
+    """
 
-    @property
-    def unhandled_prompt_behavior(self) -> str:
-        """
-        :returns: unhandled prompt behavior if set, the default is "dismiss and notify"
-        """
-        return self._caps["unhandledPromptBehavior"]
+    def __init__(self, name):
+        self.name = name
 
-    @unhandled_prompt_behavior.setter
-    def unhandled_prompt_behavior(self, behavior: str) -> None:
-        """How the driver should respond when an alert is present and the
-        command sent is not handling the alert:
-        https://w3c.github.io/webdriver/#dfn-table-of-page-load-strategies.
+    def __get__(self, obj, cls):
+        return obj._caps["unhandledPromptBehavior"]
 
-        :param behavior: behavior to use when an alert is encountered
-        """
-        if behavior in ["dismiss", "accept", "dismiss and notify", "accept and notify", "ignore"]:
-            self.set_capability("unhandledPromptBehavior", behavior)
+    def __set__(self, obj, value):
+        if value in ("dismiss", "accept", "dismiss and notify", "accept and notify", "ignore"):
+            obj.set_capability("unhandledPromptBehavior", value)
         else:
             raise ValueError(
                 "Behavior can only be one of the following: dismiss, accept, dismiss and notify, "
                 "accept and notify, ignore"
             )
 
-    @property
-    def timeouts(self) -> dict:
-        """
-        :returns: Values for implicit timeout, pageLoad timeout and script timeout if set (in milliseconds)
-        """
-        return self._caps["timeouts"]
 
-    @timeouts.setter
-    def timeouts(self, timeouts: dict) -> None:
-        """How long the driver should wait for actions to complete before
-        returning an error https://w3c.github.io/webdriver/#timeouts.
+class _Timeouts:
```

Comment:
Why do we treat PageLoadStrategy, Timeouts, Proxy and UnhandledPromptBehavior differently?

Shouldn't we be documenting all w3c capabilities this way?

Reply:
> Why do we treat PageLoadStrategy, Timeouts, Proxy and UnhandledPromptBehavior differently?
> 
> Shouldn't we be documenting all w3c capabilities this way?

We can stuff `PageLoadStrategy`, `UnHandledPromptBehavior`, `Timeouts` and `Poxy` in one single descriptor. But we need to have multiple conditional statements inside `__set__` method since the behaviour of `__set__` method is slightly different for each of the above 4 descriptors. So from code readability perspective I have just separated them out into different classes. If you observe `_BaseOptions` class, the behaviour of `__set__` method is common across attributes `browser_version`, `platform_name`, `accept_insecure_certs`, `strict_file_interactability`, `set_window_rect`. So for all the above attributes we can re-use the same `__get__` and `__set__`  method in `_BaseOptions` class. We don't have to write multiple `getters` and `setters`. Since the behaviour is same across all 5 attributes we can re-use `__get__` and `__set__` in `_BaseOptions`.

Reply:
But that means we are only getting the extra docstring details on the items that have non-default getter and/or setter behavior? Is there a way to use these descriptors instead of properties, have docstring details that can be consumed by the constructor, and have it result in less code and/or be more straightforward than what we have now with properties?

Reply:
But that means we are only getting the extra docstring details on the items that have non-default getter and/or setter behavior? Is there a way to use these descriptors instead of properties, have docstring details that can be consumed by the constructor, and have it result in less code and/or be more straightforward than what we have now with properties?